Fix #22 and add docs

tarekis opened the pull request:
Adds option `position`, which is enumerated and accepts one of two positions currently: 'bottom' and 'top'. Now the legend element will (correctly) be added to the chart container on the `created` event, and pre- or appended to the `svg`.

I also reworked the demo page, added a small documentation to it, and rephrased option descriptions to be more precise.

tarekis committed:
Clear whitespaces

tarekis committed:
Replace forEach loops with for loops

tarekis committed:
Make those for loops even faster

tarekis committed:
Encapsulate the appending logic into a handler; add enumeration parsing

tarekis committed:
Fix tests that throw errors since they exprected a synchron behaviour, which is no longer given

tarekis committed:
Clean up syntax mess

tarekis committed:
Add tests for custom positions

tarekis committed:
Add documentation and make examples beautiful

SpaceK33z commented on chartist-plugin-legend.js:
```diff
@@ -82,23 +100,36 @@
-                li.setAttribute('data-legend', i);
-                li.textContent = legend.name || legend;
-                legendElement.appendChild(li);
+            for (var legendNamesIndex = 0, length = legendNames.length; legendNamesIndex < length; legendNamesIndex++) {
```
Why did you rewrite `forEach` to `for`?

tarekis replied:
Performance is way higher, and `forEach` is plain syntax sugar. It does lots of unneeded checks on each iteration.

tarekis replied:
If jsperf wasn't down you could see the reference test [here](http://jsperf.com/for-vs-foreach/66). Let's hope jsperf v2 comes soon...

SpaceK33z replied:
I know about that test, but this is micro-optimization, and doesn't actually make a difference. I consider the advantages of `forEach` to be more important; readability, less chance of a off-by-one error, and having a separate scope. If it would be about thousands of items in the array I would consider using `for`, but for this not. Sorry.

SpaceK33z commented on chartist-plugin-legend.js:
```diff
@@ -36,6 +42,18 @@
             return a - b;
         }
 
+        // Parse input position string to enum
```
Is using an enum really necessary? Why not only allow `"bottom"` and `"top"` as a string, and throw an exception if it's not one of the two.

tarekis replied:
Not essential to the functionality, no, if you really want me to, i can change it.
Although it allows way easier intergration of possible additional positions in the future, and also the code is much cleaner IMHO; also i regard it as good practice :)

SpaceK33z replied:
Discussed this with an other team member, and we don't think using an enum is worth it. Allowing only exactly `"bottom"` or `"top"` (also no lower-case conversion) would be the most clear option.

SpaceK33z commented:
Hi, thanks for this. It's a huge PR, so it will take some time to review.

tarekis commented:
Sure thing, thanks for your time to review it!

SpaceK33z commented on index.html:
```diff
@@ -1,120 +1,649 @@
+       })
+   ]
+});</code></pre>
+      <h3>Chart with <i>clickable</i>:</h3>
```
The default value is `clickable: true`. It says here that the legend should be clickable, but you pass `clickable: false`, making it not-clickable ;).

tarekis replied:
Yeah, i actually changed the value to `false`, so actual changes for the option can be seen on the page; it would be quite irritating not to change anything, this way the viewer won't see any changes the option makes, but the description is indeed misleading. I'd change the description to:

Sets the legends clickable state; setting this value to false disables toggling graphs on legend click.

Please let me know if you can agree with that making sense.

SpaceK33z replied:
Allright, yeah it was only the text that confused me. That change would be fine.

SpaceK33z commented:
Other then the changes above, it looks good. The example page is really gorgeous now :).

SpaceK33z commented:
For a next time, it would be better to make a separate PR for each feature. It makes changes easier to review and reason about. I won't ask you to do it for this one, because that would probably be a huge pain.

tarekis commented:
You're right twice, it would be a huge pain, and i realized my error the moment i pushed the commits; gonna seperate future PR's ;)

tarekis committed:
Change documentation text for clickable

tarekis committed:
Reverted enumeration

tarekis committed:
Revert forEach to for changes

tarekis committed:
Cleaned up revertion errors

tarekis committed:
Fix syntax preference

tarekis committed:
fixed the position validator

tarekis commented:
Reverted the enumeration and loop conversion and fixed the `clickable` description.. and some things i just broke by doing just that :sweat_smile: . Please have a quick look if you deem the changes useful now.

tarekis committed:
fixed doc error

SpaceK33z commented:
Great. Thanks again!

SpaceK33z commented:
Ehmm whoops

SpaceK33z commented:
Published in `v0.5.0`.

tarekis commented:
No problem! Also, i just realized that the `index.html` belongs into the `gh-pages` branch, oops. Do you wanna change that, or should i hack together some PR?

SpaceK33z commented:
No it's fine, I'll also update the `gh-pages` branch.
